server/app/views/components/TextFormatter.java

```diff
@@ -98,16 +98,51 @@ private static String addTextSize(String markdownText) {
   }
 
   private static String addRequiredIndicator(String markdownText) {
+    // If the question ends with a list (UL or OL tag), we need to handle the required indicator
+    // differently
+    if (endsWithListTag(markdownText, "</ul>\n")) {
+      return handleRequiredQuestionsThatEndInAList(markdownText, "<ul");
+    } else if (endsWithListTag(markdownText, "</ol>\n")) {
+      return handleRequiredQuestionsThatEndInAList(markdownText, "<ol");
+    }
+
+    // If the question doesn't end with a list, add the required indicator on to the end
     int indexOfClosingTag = markdownText.lastIndexOf("</");
+    return buildStringWithRequiredIndicator(markdownText, indexOfClosingTag);
+  }
+
```

Comment:
The logic for where to render the required indicator felt confusing, so I broke it out into a bunch of mini methods. Lmk if you think this is overkill though and it would be simpler to contain the logic in fewer methods.

Reply:
Having shorter methods seems nice to me.